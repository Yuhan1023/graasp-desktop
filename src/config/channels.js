// this file needs to use module.exports as it is used both by react and electron
module.exports = {
  SAVE_SPACE_CHANNEL: 'space:save',
  GET_SPACE_CHANNEL: 'space:get',
  GET_SPACES_CHANNEL: 'spaces:get',
  DELETE_SPACE_CHANNEL: 'space:delete',
  DELETED_SPACE_CHANNEL: 'space:deleted',
  LOAD_SPACE_CHANNEL: 'space:load',
  LOADED_SPACE_CHANNEL: 'space:loaded',
  EXPORT_SPACE_CHANNEL: 'space:export',
  EXPORTED_SPACE_CHANNEL: 'space:exported',
  SHOW_OPEN_DIALOG_CHANNEL: 'show-open-dialog',
  SHOW_SAVE_DIALOG_CHANNEL: 'show-save-dialog',
  OPEN_DIALOG_PATHS_SELECTED_CHANNEL: 'open-dialog-paths-selected',
  SHOW_MESSAGE_DIALOG_CHANNEL: 'show-message-dialog',
  SAVE_DIALOG_PATH_SELECTED_CHANNEL: 'save-dialog-path-selected',
  MESSAGE_DIALOG_RESPOND_CHANNEL: 'message-dialog-respond',
};
