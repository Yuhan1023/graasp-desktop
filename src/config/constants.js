// phase item types
export const TEXT = 'text/html';
export const VIDEO = new RegExp('video/*');
export const IMAGE = new RegExp('image/*');
export const RESOURCE = 'Resource';
export const APPLICATION = 'Application';
export const IFRAME = 'application/octet-stream';
export const DEFAULT_RADIUS = 50;
export const DEFAULT_LANGUAGE = 'en';
export const DEFAULT_DEVELOPER_MODE = false;
export const DEFAULT_GEOLOCATION_ENABLED = false;
export const SHORT_ID_LENGTH = 6;
export const LONG_ID_LENGTH = 24;
export const SMART_GATEWAY_HOST = 'gateway.golabz.eu';
export const SMART_GATEWAY_QUERY_STRING_DIVIDER = '#';
export const CONTROL_TYPES = {
  BUTTON: 'BUTTON',
  SWITCH: 'SWITCH',
};
export const MIN_CARD_WIDTH = 345;
export const DEFAULT_PROTOCOL = 'https';

// math
export const BLOCK_MATH_DIV = 'p';
export const INLINE_MATH_DIV = 'span';
export const BLOCK_MATH_INDICATOR = '\\[';
export const INLINE_MATH_INDICATOR = '\\(';
export const BLOCK_MATH_REGEX = /(\\\[(.*?)\\])/g;
export const INLINE_MATH_REGEX = /(\\\((.*?)\\\))/g;
export const RADIX = 10;

export const FILTER_ALL_SPACE_ID = 'defaultid';
export const DEFAULT_AUTHENTICATION = false;
export const AUTHENTICATED = 'authenticated';

export const SYNC_ADDED = 'added';
export const SYNC_REMOVED = 'removed';
export const SYNC_MOVED = 'moved';
export const SYNC_BEFORE_MOVED = 'beforeMoved';
export const SYNC_UPDATED = 'updated';

const DIFF_COLORS = {
  [SYNC_ADDED]: '#ddffde',
  [SYNC_REMOVED]: '#fddedb',
  [SYNC_UPDATED]: '#f9edc5',
  [SYNC_MOVED]: 'lightgrey',
};

export const DIFF_STYLES = {
  [SYNC_UPDATED]: {
    background: DIFF_COLORS[SYNC_UPDATED],
  },
  [SYNC_REMOVED]: {
    background: DIFF_COLORS[SYNC_REMOVED],
  },
  [SYNC_ADDED]: {
    background: DIFF_COLORS[SYNC_ADDED],
  },
  [SYNC_MOVED]: {
    background: DIFF_COLORS[SYNC_MOVED],
  },
  [SYNC_BEFORE_MOVED]: {
    background: DIFF_COLORS[SYNC_MOVED],
    opacity: 0.5,
  },
};

export const DEFAULT_SYNC_ADVANCED_MODE = false;
export const SYNC_SPACE_PROPERTIES = [
  'name',
  'image',
  'description',
  'phases',
  'items',
];
export const SYNC_ITEM_PROPERTIES = ['name', 'description', 'content'];
