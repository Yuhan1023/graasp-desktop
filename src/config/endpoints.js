const { REACT_APP_GRAASP_HOST } = process.env;

// eslint-disable-next-line import/prefer-default-export
export const generateGetSpaceEndpoint = id =>
  `${REACT_APP_GRAASP_HOST}/spaces/${id}/download-offline`;
