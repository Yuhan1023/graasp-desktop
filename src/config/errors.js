
// same constants exist in electron.js make sure you change both
export const ERROR_ZIP_CORRUPTED = 'ERROR_ZIP_CORRUPTED';
export const ERROR_JSON_CORRUPTED = 'ERROR_JSON_CORRUPTED';
export const ERROR_SPACE_ALREADY_AVAILABLE = 'ERROR_SPACE_ALREADY_AVAILABLE';
export const ERROR_GENERAL = 'ERROR_GENERAL';
