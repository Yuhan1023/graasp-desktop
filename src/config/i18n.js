import i18n from 'i18next';
import { initReactI18next } from 'react-i18next';
import en from '../langs/en';
import fr from '../langs/fr';

i18n.use(initReactI18next).init({
  resources: {
    en_ALL: en,
    fr_ALL: fr,
  },
  fallbackLng: 'en_ALL',
  // debug only when not in production
  debug: process.env.NODE_ENV !== 'production',
  ns: ['translations'],
  defaultNS: 'translations',
  keySeparator: false,
  interpolation: {
    escapeValue: false,
    formatSeparator: ',',
  },
  react: {
    wait: true,
  },
});

const langs = {
  // bg_ALL: "български",
  // ca_ALL: "Català",
  // cs_ALL: "čeština",
  // de_ALL: "Deutsch",
  // el_ALL: "Ελληνικά",
  en_ALL: 'English',
  // es_ALL: "Español",
  // et_ALL: "Eesti",
  // fi_ALL: "Suomi",
  fr_ALL: 'Français',
  // hu_ALL: "Magyar",
  // it_ALL: "Italiano",
  // ja_ALL: "日本語",
  // ka_ALL: "ქართული",
  // lt_ALL: "lietuvių kalba",
  // lv_ALL: "Latviešu",
  // nl_ALL: "Nederlands",
  // pt_ALL: "Português",
  // ro_ALL: "Română",
  // ru_ALL: "Русский",
  // sk_ALL: "Slovenský",
  // sl_ALL: "Slovenščina",
  // sr_ALL: "српски језик",
  // sw_ALL: "Kiswahili",
  // tr_ALL: "Türkçe",
  // uk_ALL: "Українська",
  // vi_ALL: "Tiếng Việt",
  // zh_ALL: "简体中文",
  // zh_tw: "繁體中文",
};

export { langs };

export default i18n;
