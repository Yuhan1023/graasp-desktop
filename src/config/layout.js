export const MAX_APP_HEIGHT = 600;
export const MIN_APP_HEIGHT = 50;
export const DEFAULT_APP_HEIGHT = 300;
