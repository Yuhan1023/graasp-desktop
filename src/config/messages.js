// this file needs to use module.exports as it is used both by react and electron

const ERROR_DOWNLOADING_MESSAGE = 'There was a problem downloading your files';
const ERROR_LOADING_MESSAGE = 'There was a problem loading your space';
const ERROR_EXPORTING_MESSAGE = 'There was a problem exporting this space';
const SUCCESS_EXPORTING_MESSAGE = 'Space was exported successfully';
const ERROR_DELETING_MESSAGE = 'There was a problem deleting this space';
const SUCCESS_DELETING_MESSAGE = 'Space was deleted successfully';
const ERROR_ZIP_CORRUPTED_MESSAGE =
  'The archive provided is not formatted properly';
const ERROR_JSON_CORRUPTED_MESSAGE = "Space's JSON file is corrupted";
const ERROR_SPACE_ALREADY_AVAILABLE_MESSAGE =
  'A space with the same id is already available';
const SUCCESS_SPACE_LOADED_MESSAGE = 'Space was loaded successfully';
const UNEXPECTED_ERROR_MESSAGE = 'An unexpected error has occurred';
const ERROR_GETTING_SPACE_MESSAGE = 'There was an error getting that space.';
const OFFLINE_ERROR_MESSAGE = 'This functionality requires online connectivity';
const ERROR_MESSAGE_HEADER = 'Error';
const ERROR_SAVING_SPACE_MESSAGE = 'There was an error saving the space';
const SUCCESS_SAVING_MESSAGE = 'Space was saved successfully';
const SUCCESS_MESSAGE_HEADER = 'Success';

module.exports = {
  ERROR_DOWNLOADING_MESSAGE,
  ERROR_MESSAGE_HEADER,
  ERROR_SAVING_SPACE_MESSAGE,
  SUCCESS_SAVING_MESSAGE,
  SUCCESS_MESSAGE_HEADER,
  ERROR_GETTING_SPACE_MESSAGE,
  ERROR_EXPORTING_MESSAGE,
  SUCCESS_EXPORTING_MESSAGE,
  ERROR_DELETING_MESSAGE,
  SUCCESS_DELETING_MESSAGE,
  ERROR_ZIP_CORRUPTED_MESSAGE,
  ERROR_JSON_CORRUPTED_MESSAGE,
  ERROR_SPACE_ALREADY_AVAILABLE_MESSAGE,
  SUCCESS_SPACE_LOADED_MESSAGE,
  UNEXPECTED_ERROR_MESSAGE,
  OFFLINE_ERROR_MESSAGE,
  ERROR_LOADING_MESSAGE,
};
