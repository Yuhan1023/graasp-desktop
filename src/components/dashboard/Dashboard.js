import React, { Component } from 'react';
import { withStyles } from '@material-ui/core';
import _ from 'lodash';
import { connect } from 'react-redux';
import { withTranslation } from 'react-i18next';
import { withRouter } from 'react-router';
import Typography from '@material-ui/core/Typography';
import FormControl from '@material-ui/core/FormControl';
import InputLabel from '@material-ui/core/InputLabel';
import Grid from '@material-ui/core/Grid';
import Select from '@material-ui/core/Select';
import MenuItem from '@material-ui/core/MenuItem';
import PropTypes from 'prop-types';
import ActionBarChart from './ActionBarChart';
import ActionLineChart from './ActionLineChart';
import ActionTotalCount from './ActionTotalCount';
import ActionPieChart from './ActionPieChart';
import ActionEditor from './ActionEditor';
import Loader from '../common/Loader';
import Main from '../common/Main';
import { getDatabase } from '../../actions';
import { FILTER_ALL_SPACE_ID } from '../../config/constants';
import { DASHBOARD_MAIN_ID } from '../../config/selectors';

const styles = theme => ({
  dashboard: { padding: theme.spacing(3) },
  dashboardGridItem: { height: '350px' },
  drawerHeader: {
    display: 'flex',
    alignItems: 'center',
    padding: '0 8px',
    ...theme.mixins.toolbar,
    justifyContent: 'flex-end',
  },
});

export class Dashboard extends Component {
  state = {
    spaceId: FILTER_ALL_SPACE_ID,
  };

  static propTypes = {
    t: PropTypes.func.isRequired,
    classes: PropTypes.shape({
      dashboard: PropTypes.string.isRequired,
      dashboardGridItem: PropTypes.string.isRequired,
      root: PropTypes.string.isRequired,
      appBar: PropTypes.string.isRequired,
      appBarShift: PropTypes.string.isRequired,
      menuButton: PropTypes.string.isRequired,
      hide: PropTypes.string.isRequired,
      drawer: PropTypes.string.isRequired,
      drawerPaper: PropTypes.string.isRequired,
      drawerHeader: PropTypes.string.isRequired,
      content: PropTypes.string.isRequired,
      contentShift: PropTypes.string.isRequired,
      developer: PropTypes.string.isRequired,
      screenTitle: PropTypes.string.isRequired,
    }).isRequired,
    theme: PropTypes.shape({
      direction: PropTypes.string.isRequired,
    }).isRequired,
    history: PropTypes.shape({
      replace: PropTypes.func.isRequired,
    }).isRequired,
    i18n: PropTypes.shape({
      changeLanguage: PropTypes.func.isRequired,
    }).isRequired,
    database: PropTypes.shape({
      user: PropTypes.object,
      spaces: PropTypes.arrayOf(PropTypes.object),
      actions: PropTypes.arrayOf(PropTypes.object),
    }),
    dispatchGetDatabase: PropTypes.func.isRequired,
  };

  static defaultProps = {
    database: {},
  };

  componentDidMount() {
    const { dispatchGetDatabase } = this.props;
    dispatchGetDatabase();
  }

  handleSpaceChange = event => {
    this.setState({ spaceId: event.target.value });
  };

  renderSpaceFilter = () => {
    const { database, t } = this.props;
    const { spaceId } = this.state;

    if (!database || _.isEmpty(database)) {
      return <Loader />;
    }

    return (
      <FormControl variant="outlined" fullWidth>
        <InputLabel>{t('Filter by Space')}</InputLabel>
        <Select
          label="Filter by Space"
          value={spaceId}
          onChange={this.handleSpaceChange}
        >
          <MenuItem value={FILTER_ALL_SPACE_ID}>
            <em>{t('All Spaces')}</em>
          </MenuItem>
          {database.spaces.map(space => (
            <MenuItem value={space.id}>{space.name}</MenuItem>
          ))}
        </Select>
      </FormControl>
    );
  };

  renderActionWidgets = () => {
    const { database, t, classes } = this.props;
    const { spaceId } = this.state;

    let filteredActions = database.actions;
    if (spaceId !== FILTER_ALL_SPACE_ID) {
      filteredActions = filteredActions.filter(
        ({ spaceId: id }) => id === spaceId
      );
    }

    if (_.isEmpty(filteredActions)) {
      return <p>{t('No action has been recorded.')}</p>;
    }

    return (
      <Grid
        style={{ display: 'flex' }}
        spacing={5}
        container
        justify="center"
        alignItems="center"
      >
        <Grid item xs={12} sm={6} className={classes.dashboardGridItem}>
          <ActionBarChart spaces={database.spaces} actions={filteredActions} />
        </Grid>
        <Grid item xs={12} sm={6} className={classes.dashboardGridItem}>
          <ActionLineChart actions={filteredActions} />
        </Grid>
        <Grid item xs={12} sm={6} className={classes.dashboardGridItem}>
          <ActionPieChart actions={filteredActions} />
        </Grid>
        <Grid item xs={12} sm={6} className={classes.dashboardGridItem}>
          <ActionTotalCount actions={filteredActions} />
        </Grid>
      </Grid>
    );
  };

  render() {
    const { classes, t, database } = this.props;
    const { spaceId } = this.state;

    if (!database) {
      return <Loader />;
    }

    if (_.isEmpty(database)) {
      return <p>{t('The database is empty.')}</p>;
    }

    return (
      <Main id={DASHBOARD_MAIN_ID}>
        <div className={classes.dashboard}>
          <Grid
            style={{ display: 'flex' }}
            container
            justify="center"
            alignItems="center"
            spacing={3}
          >
            <Grid item xs={12} sm={9}>
              <Typography variant="h4" className={classes.screenTitle}>
                {t('Dashboard')}
              </Typography>
            </Grid>
            <Grid item xs={12} sm={3}>
              {this.renderSpaceFilter()}
            </Grid>
          </Grid>

          {this.renderActionWidgets()}

          <ActionEditor spaceId={spaceId} />
        </div>
      </Main>
    );
  }
}

const mapStateToProps = ({ Developer }) => ({
  database: Developer.get('database'),
});

const mapDispatchToProps = {
  dispatchGetDatabase: getDatabase,
};

const ConnectedComponent = connect(
  mapStateToProps,
  mapDispatchToProps
)(Dashboard);

const StyledComponent = withStyles(styles, { withTheme: true })(
  ConnectedComponent
);

const TranslatedComponent = withTranslation()(StyledComponent);

export default withRouter(TranslatedComponent);
