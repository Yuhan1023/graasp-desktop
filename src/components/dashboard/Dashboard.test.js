import React from 'react';
import { shallow } from 'enzyme';
import Typography from '@material-ui/core/Typography';
import FormControl from '@material-ui/core/FormControl';
import MenuItem from '@material-ui/core/MenuItem';
import { Dashboard } from './Dashboard';
import { DEFAULT_USER_MODE, USER_MODES } from '../../config/constants';
import {
  DASHBOARD_ACTION_EDITOR_ID,
  DASHBOARD_BAR_CHART_ID,
  DASHBOARD_USER_FILTER_ID,
  DASHBOARD_LINE_CHART_ID,
  DASHBOARD_NO_ACTION_MESSAGE_ID,
  DASHBOARD_PIE_CHART_ID,
  DASHBOARD_SPACE_FILTER_ID,
  DASHBOARD_TOTAL_COUNT_ID,
} from '../../config/selectors';

const sampleUserId = '123456';
const emptyUserId = 'empty';
const sampleUsers = [
  { id: sampleUserId, username: 'bob' },
  { id: 'randomId', username: 'anna' },
  { id: emptyUserId, username: 'cedric' },
];
const sampleSpaces = [{ id: 'spaceId', name: 'spaceName' }];
const sampleActions = [
  { id: 'actionId', user: sampleUserId, spaceId: 'spaceId' },
];
const sampleDatabase = {
  actions: sampleActions,
  users: sampleUsers,
  spaces: sampleSpaces,
};

const createDashboardProps = ({
  userId,
  userMode = DEFAULT_USER_MODE,
  users,
  actions,
  spaces,
} = {}) => {
  return {
    t: jest.fn((text) => text),
    classes: {
      dashboard: '',
      dashboardGridItem: '',
      root: '',
      appBar: '',
      appBarShift: '',
      menuButton: '',
      hide: '',
      drawer: '',
      drawerPaper: '',
      drawerHeader: '',
      content: '',
      contentShift: '',
      developer: '',
      screenTitle: '',
    },
    theme: {
      direction: '',
    },
    history: {
      replace: jest.fn(),
    },
    i18n: {
      changeLanguage: jest.fn(),
    },
    database: {
      spaces,
      users,
      actions,
    },
    dispatchGetDatabase: jest.fn(),
    userMode,
    userId,
  };
};

describe('<Dashboard />', () => {
  let wrapper;

  describe('default props', () => {
    beforeAll(() => {
      const props = createDashboardProps();
      // eslint-disable-next-line react/jsx-props-no-spreading
      wrapper = shallow(<Dashboard {...props} />);
    });

    it('renders correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('no actions renders message', () => {
      const typography = wrapper
        .find(Typography)
        .find(`#${DASHBOARD_NO_ACTION_MESSAGE_ID}`);
      expect(typography).toHaveLength(1);
      expect(typography.contains('No action has been recorded.')).toBeTruthy();
    });
  });

  describe('actions is defined', () => {
    describe('userMode = student', () => {
      it('displays no actions if no action exists for given userId', () => {
        const props = createDashboardProps({
          userId: emptyUserId,
          ...sampleDatabase,
        });
        // eslint-disable-next-line react/jsx-props-no-spreading
        wrapper = shallow(<Dashboard {...props} />);
        const typography = wrapper
          .find(Typography)
          .find(`#${DASHBOARD_NO_ACTION_MESSAGE_ID}`);
        expect(typography).toHaveLength(1);
        expect(
          typography.contains('No action has been recorded.')
        ).toBeTruthy();
      });

      it('displays 4 visualizations widgets, editor, space filter for given userId', () => {
        const props = createDashboardProps({
          userId: sampleUserId,
          ...sampleDatabase,
        });
        // eslint-disable-next-line react/jsx-props-no-spreading
        wrapper = shallow(<Dashboard {...props} />);
        const barChart = wrapper.find(`#${DASHBOARD_BAR_CHART_ID}`);
        expect(barChart).toHaveLength(1);
        expect(barChart).toHaveLength(1);
        const lineChart = wrapper.find(`#${DASHBOARD_LINE_CHART_ID}`);
        expect(lineChart).toHaveLength(1);
        const pieChart = wrapper.find(`#${DASHBOARD_PIE_CHART_ID}`);
        expect(pieChart).toHaveLength(1);
        const editor = wrapper.find(`#${DASHBOARD_ACTION_EDITOR_ID}`);
        expect(editor).toHaveLength(1);
        const totalCount = wrapper.find(`#${DASHBOARD_TOTAL_COUNT_ID}`);
        expect(totalCount).toHaveLength(1);

        const spaceFilter = wrapper.find(`#${DASHBOARD_SPACE_FILTER_ID}`);
        expect(spaceFilter).toHaveLength(1);

        const userFilter = wrapper.find(`#${DASHBOARD_USER_FILTER_ID}`);
        expect(userFilter).toHaveLength(0);
      });
    });

    describe(`userMode = teacher`, () => {
      beforeAll(() => {
        const props = createDashboardProps({
          userId: sampleUserId,
          userMode: USER_MODES.TEACHER,
          ...sampleDatabase,
        });
        // eslint-disable-next-line react/jsx-props-no-spreading
        wrapper = shallow(<Dashboard {...props} />);
      });

      it('displays user and space filters', () => {
        const spaceFilter = wrapper.find(`#${DASHBOARD_SPACE_FILTER_ID}`);
        expect(spaceFilter).toHaveLength(1);

        const formControl = wrapper
          .find(FormControl)
          .find(`#${DASHBOARD_USER_FILTER_ID}`);
        const menuItems = formControl.find(MenuItem).map((item) => item.text());

        expect(menuItems).toContain('All Users');
        // eslint-disable-next-line no-restricted-syntax
        for (const user of sampleUsers) {
          expect(menuItems).toContain(user.username);
        }

        expect(menuItems.length).toBe(sampleUsers.length + 1);
      });
    });
  });
});
