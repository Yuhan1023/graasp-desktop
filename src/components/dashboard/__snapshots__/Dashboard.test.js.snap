// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dashboard /> default props renders correctly 1`] = `
<withI18nextTranslation(WithStyles(Connect(Main)))
  id="dashboardMain"
>
  <div
    className=""
  >
    <WithStyles(ForwardRef(Grid))
      alignItems="center"
      container={true}
      justify="center"
      spacing={3}
      style={
        Object {
          "display": "flex",
        }
      }
    >
      <WithStyles(ForwardRef(Grid))
        item={true}
        sm={6}
        xs={12}
      >
        <WithStyles(ForwardRef(Typography))
          className=""
          variant="h4"
        >
          Dashboard
        </WithStyles(ForwardRef(Typography))>
      </WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(Grid))
        item={true}
        sm={3}
        xs={12}
      />
      <WithStyles(ForwardRef(Grid))
        item={true}
        sm={3}
        xs={12}
      >
        <WithStyles(ForwardRef(FormControl))
          fullWidth={true}
          id="dashboardSpaceFilter"
          variant="outlined"
        >
          <WithStyles(ForwardRef(InputLabel))>
            Filter by Space
          </WithStyles(ForwardRef(InputLabel))>
          <WithStyles(ForwardRef(Select))
            label="Filter by Space"
            onChange={[Function]}
            value="allSpacesSelectId"
          >
            <WithStyles(ForwardRef(MenuItem))
              value="allSpacesSelectId"
            >
              <em>
                All Spaces
              </em>
            </WithStyles(ForwardRef(MenuItem))>
          </WithStyles(ForwardRef(Select))>
        </WithStyles(ForwardRef(FormControl))>
      </WithStyles(ForwardRef(Grid))>
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Typography))
      id="dashboardNoActionMessage"
    >
      No action has been recorded.
    </WithStyles(ForwardRef(Typography))>
  </div>
</withI18nextTranslation(WithStyles(Connect(Main)))>
`;
